DependencyInjection/AclExtension.php

```diff
@@ -0,0 +1,117 @@
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ */
+class AclExtension extends Extension implements PrependExtensionInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $mainConfig = $this->getConfiguration($configs, $container);
+        $config = $this->processConfiguration($mainConfig, $configs);
+
+        $loader = new XmlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('security_acl.xml');
+
+        if (class_exists(Application::class)) {
+            $loader->load('console.xml');
+        }
+
+        if (isset($config['cache']['id'])) {
+            $container->setAlias('security.acl.cache', $config['cache']['id']);
+        }
+        $container->getDefinition('security.acl.voter.basic_permissions')->addArgument($config['voter']['allow_if_object_identity_unavailable']);
+
+        // custom ACL provider
+        if (isset($config['provider'])) {
+            $container->setAlias('security.acl.provider', $config['provider']);
+
+            return;
+        }
+
+        // no provider configured
+        $loader->load('security_acl_dbal.xml');
+
+        if (null !== $config['connection']) {
+            $container->setAlias('security.acl.dbal.connection', sprintf('doctrine.dbal.%s_connection', $config['connection']));
+        }
+
+        $container
+            ->getDefinition('security.acl.dbal.schema_listener')
+            ->addTag('doctrine.event_listener', array(
+                'connection' => $config['connection'],
+                'event' => 'postGenerateSchema',
+                'lazy' => true,
+            ))
+        ;
+
+        $container->getDefinition('security.acl.cache.doctrine')->addArgument($config['cache']['prefix']);
+
+        $container->setParameter('security.acl.dbal.class_table_name', $config['tables']['class']);
+        $container->setParameter('security.acl.dbal.entry_table_name', $config['tables']['entry']);
+        $container->setParameter('security.acl.dbal.oid_table_name', $config['tables']['object_identity']);
+        $container->setParameter('security.acl.dbal.oid_ancestors_table_name', $config['tables']['object_identity_ancestors']);
+        $container->setParameter('security.acl.dbal.sid_table_name', $config['tables']['security_identity']);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getXsdValidationBasePath()
+    {
+        return __DIR__.'/../Resources/config/schema';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getNamespace()
+    {
+        return 'http://symfony.com/schema/dic/acl';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prepend(ContainerBuilder $container)
+    {
+        $securityConfig = $container->getExtensionConfig('security');
+        $aclConfig = $container->getExtensionConfig('acl');
+
+        if (empty($securityConfig['acl'])) {
```

Comment:
that's broken. `$securityConfig` is an array of configuration arrays, as arrays are not merged yet at this point (and are never merged inside the container extension config btw).

You need to loop over all of them and check each of them to prepend them (be careful about the order, to avoid reversing the array).